evm/src/contracts/CircleIntegration/Logic.sol

```diff
@@ -0,0 +1,240 @@
+
+abstract contract Logic is ICircleIntegration, Governance {
+    using Utils for address;
+    using SafeERC20 for IERC20;
+    using WormholeCctpMessages for *;
+
+    /// @inheritdoc ICircleIntegration
+    function transferTokensWithPayload(
+        TransferParameters calldata transferParams,
+        uint32 wormholeNonce,
+        bytes calldata payload
+    ) public payable returns (uint64 wormholeSequence) {
+        // Is the foreign Wormhole Circle Integration registered?
+        bytes32 destinationCaller = getRegisteredEmitters()[transferParams.targetChain];
+        require(destinationCaller != 0, "target contract not registered");
+
+        // Deposit tokens into this contract to prepare for burning.
+        IERC20(transferParams.token).safeTransferFrom(
+            msg.sender, address(this), transferParams.amount
+        );
+
+        // Approve the Token Messenger to spend tokens.
+        setTokenMessengerApproval(transferParams.token, transferParams.amount);
+
+        // Invoke Token Messenger to burn tokens and emit a CCTP token burn message.
+        (wormholeSequence,) = burnAndPublish(
+            destinationCaller,
+            getChainToDomain()[transferParams.targetChain],
+            transferParams.token,
+            transferParams.amount,
+            transferParams.mintRecipient,
+            wormholeNonce,
+            payload,
+            msg.value
+        );
+    }
+
+    /// @inheritdoc ICircleIntegration
+    function redeemTokensWithPayload(RedeemParameters calldata params)
+        public
+        returns (DepositWithPayload memory deposit)
+    {
+        // This check prevents this contract existing on this network's potential fork, where it was
+        // not freshly deployed. This is a safety measure to prevent replay attacks on the forked
+        // network.
+        require(evmChain() == block.chainid, "invalid evm chain");
+
+        // Verify the VAA and mint tokens. Set the deposit struct with WormholeCctpTokenMessenger's
+        // return values.
+        IWormhole.VM memory vaa;
+        (
+            vaa,
+            deposit.token,
+            deposit.amount,
+            deposit.sourceDomain,
+            deposit.targetDomain,
+            deposit.nonce,
+            deposit.fromAddress,
+            deposit.mintRecipient,
+            deposit.payload
+        ) = verifyVaaAndMintLegacy(
+            params.encodedCctpMessage, params.cctpAttestation, params.encodedVaa
+        );
+
+        // Confirm that the caller is the `mintRecipient` to ensure atomic execution.
+        require(
+            msg.sender.toUniversalAddress() == deposit.mintRecipient, "caller must be mintRecipient"
+        );
+
+        // If this VAA does not come from a registered Wormhole Circle Integration contract, revert.
+        require(
+            vaa.emitterAddress != 0
+                && vaa.emitterAddress == getRegisteredEmitters()[vaa.emitterChainId],
+            "unknown emitter"
+        );
```

Comment:
What do you think about moving this logic (in another form) to the `WormholeCctpTokenMessenger` contract? Without this check integrators can shoot themselves in the foot.

I was thinking we could add this check somewhere in `_parseAndVerifyVAA` and allow the integrator to pass in an address to check against (i.e. always compare the VAA emitter address to some other address)

Reply:
Added `requireEmitter` and `requireEmitterLegacy` here: https://github.com/wormhole-foundation/wormhole-circle-integration/pull/48/commits/e727f5349be794decc3b59d4864afd9986001707